### PR TITLE
Add overlay service to dispatch and handle requests

### PR DESCRIPTION
**Pull request by jacobkaufmann**
Related to #135. Move message dispatch and handling into `OverlayService`. The `Overlay` communicates with the associated `OverlayService` via a channel.

Motivation: Autonomous routing table management requires a separate task where the task must have a mutable reference to the overlay structure. We create a new overlay structure, `OverlayService`, and communication between this structure and `Overlay` occurs over a channel, which does not require a reference.

### CODE REVIEW

**Comment by njgheorghita** [comment]
@jacobkaufmann Yeah, it's a bit large but also sometimes that's unavoidable. It also seem like it covers some of the same territory as #163, #164 - so maybe wait until those land (which should be soon) and then either rebase off them, or leave your changes here if you feel like it's a more suitable error structure (which very well might be the case).

**Review by njgheorghita — APPROVED** [review]
Looks great! Nice work, just one question, so it's mentioned that the overlay routing table maintenance / liveness tasks are done autonomously - which I interpret as handled within `discv5` - so my understanding here is that the maintenance / liveness tasks are done over the base discv5 protocol? This seems fine for `PING`, but are `FINDNODES` lookup/bootstrapping messages part of the autonomous management at all? If so, and those are going out over the base discv5 protocol, that would be corrupt the routing table

File: ethportal-peertest/src/main.rs
Comment (jacobkaufmann) [reply ethportal-peertest/src/main.rs:38]: Ah yes, good catch, this was a miss by me.

File: trin-history/src/events.rs
Comment (jacobkaufmann) [reply trin-history/src/events.rs:32]: Yeah, I agree. Initially I changed everything to be `Err(error)` and not `Err(e)`, but I think I switched back to avoid unrelated changes. Looks like I missed some of those. I'll go ahead and update the `Err(e)` to `Err(error)` in parts of the code that this PR touches. I agree that we should standardize a convention in a later PR.

File: ethportal-peertest/src/events.rs
Comment (jacobkaufmann) [reply ethportal-peertest/src/events.rs:52]: I agree here. I just kept the pre-existing message, but I can go ahead and update that.

**Comment by jacobkaufmann** [comment]
> Looks great! Nice work, just one question, so it's mentioned that the overlay routing table maintenance / liveness tasks are done autonomously - which I interpret as handled within `discv5` - so my understanding here is that the maintenance / liveness tasks are done over the base discv5 protocol? This seems fine for `PING`, but are `FINDNODES` lookup/bootstrapping messages part of the autonomous management at all? If so, and those are going out over the base discv5 protocol, that would be corrupt the routing table

@njgheorghita I think you are referring to the comment above `OverlayService::spawn`? So maybe the remark about "autonomous" processes here is premature, because those processes are not implemented in this PR. If you see #135 in the `overlay/service.rs` file, the main loop performs pings based on the ping queue, (#139 introduces the necessary data structure for the ping queue), and the loop also performs some maintenance tasks on the routing table for previously applied pending entries.

These processes are considered "autonomous" because we do not need to invoke a method on `OverlayService` for them to occur. We just have to spawn the service. Also, these processes do not occur over `discv5` `PING` or `FINDNODES` messages. The ping messages that are sent are overlay `PING` messages, and the node updates are done via overlay `FINDNODES` messages that contain the latest ENR.

I hope this clarifies things. If the autonomous comment is confusing without the implementation of those processes, then I'm happy to remove it. And if anything else deserves some further elaboration, then I can add additional comments.

**Comment by njgheorghita** [comment]
@jacobkaufmann Ahh cool, yeah that makes perfect sense, thanks!

File: ethportal-peertest/src/events.rs
Comment (ogenev) [inline ethportal-peertest/src/events.rs:47]: FYI this match will not work with the latest ProtocolId update in https://github.com/ethereum/trin/pull/162. I'll create an issue to fix this in a separate PR.

**Comment by jacobkaufmann** [comment]
I think I've addressed all the comments. Waiting on #176, which @ogenev created to address the comment above, to land before I merge to ensure there are no conflicts.

**Pull request merged**
